### PR TITLE
ZSTD compressor

### PR DESCRIPTION
**Pull request by martint**

### CODE REVIEW

File: src/test/java/io/airlift/compress/zstd/TestZstd.java [hunk 1]
```diff
@@ -90,4 +90,24 @@ public void testInvalidSequenceOffset()
 
         getDecompressor().decompress(compressed, 0, compressed.length, output, 0, output.length);
     }
+
+    @Test
+    public void testSmallLiteralsAfterIncompressibleLiterals()
```
Comment (martint) [inline src/test/java/io/airlift/compress/zstd/TestZstd.java]: TODO: this test should move to the commit that adds huffman

File: src/main/java/io/airlift/compress/zstd/SequenceStore.java [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+        sequenceCount++;
+    }
+
+    public void reset()
+    {
+        literalsLength = 0;
+        sequenceCount = 0;
+        longLengthField = null;
+
+//        start = 0; // TODO for debugging
```
Comment (martint) [inline src/main/java/io/airlift/compress/zstd/SequenceStore.java]: TODO: remove

File: src/main/java/io/airlift/compress/zstd/SequenceStore.java [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+    {
+        UNSAFE.copyMemory(inputBase, inputAddress, literalsBuffer, ARRAY_BYTE_BASE_OFFSET + literalsLength, inputSize);
+        literalsLength += inputSize;
+    }
+
+    public void storeSequence(Object literalBase, long literalAddress, int literalLength, int offsetCode, int matchLengthBase)
+    {
+        // TODO: store codes directly?
+
+//        if (start == 0) {
```
Comment (martint) [inline src/main/java/io/airlift/compress/zstd/SequenceStore.java]: TODO: remove commented out alternatives

**Comment by wgtmac** [comment]
Hi @martint @dain do you have an ETA of this PR?

File: src/main/java/io/airlift/compress/zstd/ZstdFrameCompressor.java [hunk 4]
```diff
@@ -244,6 +261,151 @@ private static int compressBlock(Object inputBase, long inputAddress, int inputS
...
+
+        int[] counts = new int[MAX_SYMBOL_COUNT]; // TODO: preallocate
+        Histogram.count(literals, literalsSize, counts);
+        int maxSymbol = Histogram.findMaxSymbol(counts, MAX_SYMBOL);
+        int largestCount = Histogram.findLargestCount(counts, maxSymbol);
+
+        long literalsAddress = ARRAY_BYTE_BASE_OFFSET;
+        if (largestCount == literalsSize) {
+            // all bytes in input are equal
+            UNSAFE.putByte(outputBase, outputAddress + headerSize, UNSAFE.getByte(literals, literalsAddress));
```
Comment (dain) [inline src/main/java/io/airlift/compress/zstd/ZstdFrameCompressor.java]: Is this needed?

**Comment by findepi** [comment]
🚀

**Pull request merged**